# Tugas Praktikum

Ubah jumlah neuron di hidden layer (misal: 256 dan 128).

In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(128, activation='relu'), # Hidden layer 1
    Dense(64, activation='relu'),  # Hidden layer 2
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8783 - loss: 0.4134 - val_accuracy: 0.9673 - val_loss: 0.1058
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 6ms/step - accuracy: 0.9686 - loss: 0.1045 - val_accuracy: 0.9705 - val_loss: 0.0959
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9787 - loss: 0.0688 - val_accuracy: 0.9741 - val_loss: 0.0851
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.9845 - loss: 0.0488 - val_accuracy: 0.9739 - val_loss: 0.0827
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 17s 6ms/step - accuracy: 0.9880 - loss: 0.0384 - val_accuracy: 0.9754 - val_loss: 0.0788
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9893 - loss: 0.0312 - val_accuracy: 0.9786 - val_loss: 0.0739
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9919 - loss: 0.0243 - val_accuracy: 0.9792 - val_loss: 0.0768
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 6ms/step - accuracy: 0.9931 - loss: 0

Tambahkan satu hidden layer lagi.



In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(128, activation='relu'), # Hidden layer 1
    Dense(64, activation='relu'),  # Hidden layer 2
    Dense(32, activation='relu'),
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.8633 - loss: 0.4526 - val_accuracy: 0.9585 - val_loss: 0.1267
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9669 - loss: 0.1093 - val_accuracy: 0.9700 - val_loss: 0.0966
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.9759 - loss: 0.0742 - val_accuracy: 0.9730 - val_loss: 0.0859
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9828 - loss: 0.0534 - val_accuracy: 0.9715 - val_loss: 0.0922
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9858 - loss: 0.0442 - val_accuracy: 0.9770 - val_loss: 0.0814
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 19s 6ms/step - accuracy: 0.9889 - loss: 0.0347 - val_accuracy: 0.9760 - val_loss: 0.0907
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 6ms/step - accuracy: 0.9900 - loss: 0.0299 - val_accuracy: 0.9769 - val_loss: 0.0867
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9920 - loss: 0.

Bandingkan akurasi dan waktu pelatihan.

1. Sebelum Menambah Hidden Layer

Akurasi: 0.9775
Model dengan dua hidden layer (128 → 64) sudah cukup optimal.
Loss dan akurasi stabil, tidak terjadi overfitting yang signifikan.

2. Setelah Menambah Satu Hidden Layer

(Hidden layer menjadi: 128 → 64 → 32)
Akurasi: 0.9759

Hasilnya justru sedikit menurun dibandingkan model awal.

In [ ]:
import tensorflow as tf
import time
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Fungsi untuk training + hitung waktu
def train_model(activation):
    print(f"\n===== Model dengan aktivasi {activation.upper()} =====")

    model = Sequential([
        Flatten(input_shape=(28, 28)),
        Dense(128, activation=activation),
        Dense(64, activation=activation),
        Dense(32, activation=activation),
        Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

    start = time.time()
    history = model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=0)
    end = time.time()

    loss, acc = model.evaluate(X_test, y_test, verbose=0)

    print(f"Akurasi : {acc:.4f}")
    print(f"Waktu   : {end - start:.2f} detik")

    return acc, end - start

# 2. Eksperimen ReLU
relu_acc, relu_time = train_model('relu')

# 3. Eksperimen Sigmoid
sig_acc, sig_time = train_model('sigmoid')

# 4. Ringkasan hasil
print("\n================ RINGKASAN PERBANDINGAN ================")
print(f"ReLU    → Akurasi: {relu_acc:.4f}, Waktu: {relu_time:.2f} detik")
print(f"Sigmoid → Akurasi: {sig_acc:.4f}, Waktu: {sig_time:.2f} detik")



===== Model dengan aktivasi RELU =====
Akurasi : 0.9768
Waktu   : 90.61 detik

===== Model dengan aktivasi SIGMOID =====
Akurasi : 0.9741
Waktu   : 89.16 detik

================ RINGKASAN PERBANDINGAN ================
ReLU    → Akurasi: 0.9768, Waktu: 90.61 detik
Sigmoid → Akurasi: 0.9741, Waktu: 89.16 detik
